android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/MenuPositioningData.kt

```diff
@@ -0,0 +1,270 @@
+    if (!hasViewInitializedCorrectly) {
+        Log.log(
+            priority = Log.Priority.ERROR,
+            tag = "mozac-popupwindow",
+            message = "containerViewHeight: $containerViewHeight must be more than 0," +
+                "containerViewWidth: $containerViewWidth must be more than 0," +
+                "recyclerViewHeight: $recyclerViewHeight must be more than 0 &" +
+                "recyclerViewItemCount: $recyclerViewItemCount must be more than 0.",
+        )
+        return null
+    }
+
+    val menuHorizontalPadding = containerViewWidth - recyclerView.measuredWidth
+    val menuVerticalPadding = containerViewHeight - recyclerView.measuredHeight
+    var horizontalOffset = style?.horizontalOffset ?: 0
+    var verticalOffset = style?.verticalOffset ?: 0
+
+    // Elevation creates some padding between the menu and its container, so that the start corner
+    // of the menu doesn't match the corner of the anchor view. If the user wants the menu to hide
+    // the anchor completely, we have to adjust the position of the menu to compensate for the inner
+    // padding of the menu container.
+    if (style?.completelyOverlap == true) {
+        horizontalOffset -= menuHorizontalPadding / 2
+        verticalOffset -= menuVerticalPadding / 2
+    }
+
+    // The menu height might be adjusted: if there is not enough space to show all the items,
+    // it will crop the last visible item in half, to give the user a hint that it is scrollable.
+    containerViewHeight = calculateContainerHeight(
+        recyclerViewHeight,
+        recyclerViewItemCount,
+        containerViewHeight,
+        style?.verticalOffset ?: 0,
+        anchor,
+    )
+
+    val (topSpace, bottomSpace) = getMaxAvailableHeightToTopAndBottom(anchor)
+    val (leftSpace, rightSpace) = getMaxAvailableWidthToLeftAndRight(anchor)
+
+    val fitsUp = topSpace + anchor.height >= containerViewHeight
+    val fitsDown = bottomSpace + anchor.height >= containerViewHeight
+    val fitsRight = rightSpace + anchor.width >= containerViewWidth
+    val fitsLeft = leftSpace + anchor.width >= containerViewWidth
+
+    val notEnoughHorizontalSpace = !fitsRight && !fitsLeft
+    val fitsBothHorizontalDirections = fitsRight && fitsLeft
+    val drawingLeft = if (notEnoughHorizontalSpace || fitsBothHorizontalDirections) {
+        anchor.isRTL
+    } else {
+        !fitsRight
+    }
+
+    val anchorPosition = IntArray(2)
+    anchor.getLocationInWindow(anchorPosition)
+    var (x, y) = anchorPosition
+
+    // Position the menu above the anchor if the orientation is UP and there is enough space.
+    if (orientation == Orientation.UP && fitsUp) {
+        y -= containerViewHeight - anchor.height
+        verticalOffset = -verticalOffset
+    }
+
+    if (drawingLeft) {
+        x -= containerViewWidth - anchor.width
+        horizontalOffset = -horizontalOffset
+    }
+
+    return MenuPositioningData(
+        x = x + horizontalOffset,
```

Comment:
```suggestion
        x += + horizontalOffset,
```

Reply:
I was confused about this as well but I think Mike has this right though.  He's not updating `x` or `y` here.  It's just unfortunate that the parameter name is the same.

Reply:
Ah good catch, so it is! 🙃

Reply:
fair comment, it does look confusing! changed the naming